Fix for issue #1117, temporary fix for set_ttl test on Windows, this …

dtacalau opened the pull request:
…test was crashing on Windows because, on Windows calling setsockopt on a

not connected socket is undefined behavior. More details can be found in issue #1117.

Signed-off-by: Daniel Tacalau <dst4096@gmail.com>

Thomasdezeeuw commented:
This change looks good to me, two questions:
* Does this apply to more methods on `TcpStream`?
* Can you add some documentation to `TcpStream::set_ttl` (and perhaps other methods) that it's required to receive an before using the method?

dtacalau commented:
> * Does this apply to more methods on `TcpStream`?

@Thomasdezeeuw you mean other TcpStream methods which are using setsockopt? It so, there's only set_nodelay which suffers from the same issue, but its test has been already fixed.

> * Can you add some documentation to `TcpStream::set_ttl` (and perhaps other methods) that it's required to receive an before using the method?

Absolutely, ATM there are two affected: set_ttl and set_nodelay, others?

Thomasdezeeuw commented:
If the problem is calling `setsockopt` on an unconnected socket then I think only `set_ttl` and `set_nodelay` are a problem. So only those two methods would need docs.

But does that same hold for `getsockopt`? Then it would also apply to `ttl` and `nodelay` (the getter methods).

dtacalau commented:
> But does that same hold for `getsockopt`? Then it would also apply to `ttl` and `nodelay` (the getter methods).

Could be, will check all 4 of them.

dtacalau commented:
Latest changes:
- on Windows, all 4 methods: set ttl, get ttl, set nodelay, get nodelay require a connected stream to work. I've added tests for all of them, also added documentation notes.
- added set/get ttl tests for TcpListener;
- added set/get ttl tests for UdpSocket;

Thomasdezeeuw commented on src/net/tcp_stream.rs:
```diff
@@ -105,6 +105,9 @@ impl TcpStream {
     /// sufficient amount to send out, thereby avoiding the frequent sending of
     /// small packets.
+    ///
+    /// Note On Windows: Make sure the stream is connected before calling this method.
```
Can you changes this to:

```rust
    /// # Notes
    /// 
    /// On Windows make sure the stream is connected before calling this method,
    /// by receiving an (writable) event. Trying to set `nodelay` on an
    /// unconnected `TcpStream` is undefined behaviour.
```

Same for the others.

dtacalau replied:
Done, reworded.

Thomasdezeeuw replied:
You missed the Notes header.

dtacalau replied:
Sorry for that, I've added it.

Thomasdezeeuw commented on tests/udp_socket.rs:
```diff
@@ -111,6 +111,26 @@ fn smoke_test_unconnected_udp_socket(socket1: UdpSocket, socket2: UdpSocket) {
+    socket1.ttl().expect("unable to get TTL for UDP socket");
+}
+
+
```
Double empty line.

dtacalau replied:
Done, removed.

dtacalau commented:
> Can you rebase on master and fix the last couple things, then we can merge this.

Done
